File: public/dsa/searching/insert_position_in_sorted_array/left_binary_search/insert_position_in_sorted_array.cpp
#include <bits/stdc++.h>
using namespace std;

int left_binary_search(const vector<int> &nums, int target) {
    int left = 0, right = nums.size() - 1;
    while (left < right) {
        int mid = (left + right) / 2;
        if (nums[mid] < target) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    return left;
}

int getInsertPosition(vector<int> &arr, int key) {
    return left_binary_search(arr, key);
}

// tests-start
int main() {
    vector<int> arr = {1, 2, 3, 5};
    int key = 4;
    cout << getInsertPosition(arr, key) << endl;
    return 0;
}
// tests-end

/*output
3
*/

File: public/dsa/searching/insert_position_in_sorted_array/left_binary_search/insert_position_in_sorted_array.py


from typing import List


def left_binary_search(nums: List[int], target: int) -> int:
    left = 0
    right = len(nums)
    while left < right:
        mid = (left + right) // 2
        if nums[mid] < target:
            left = mid + 1
        else:
            right = mid
    return left


class Solution:
    def getInsertPosition(self, arr: List[int], key: int) -> int:
        return left_binary_search(arr, key)


# tests-start
arr = [1, 2, 3, 5]
key = 4
sol = Solution()
print(sol.getInsertPosition(arr, key))
# tests-end


'''output
3
'''
